src/main/java/io/prometheus/cloudwatch/CachingDimensionSource.java

```diff
@@ -0,0 +1,132 @@
+package io.prometheus.cloudwatch;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+final class CachingDimensionSource implements DimensionSource {
+
+  private static final int MAX_ENTRIES = 500;
+  private final DimensionSource delegate;
+  private final Cache<DimensionCacheKey, DimensionData> cache;
+
+  CachingDimensionSource(DimensionSource delegate, Cache<DimensionCacheKey, DimensionData> cache) {
+    this.delegate = delegate;
+    this.cache = cache;
+  }
+
+  /**
+   * Create a new DimensionSource that will cache the results from another {@link DimensionSource}
+   *
+   * @param source
+   * @param config - config used to configure the expiry (ttl) for entries in the cache
+   * @return a new CachingDimensionSource
+   */
+  static DimensionSource create(DimensionSource source, DimensionCacheConfig config) {
```

Comment:
What would become impossible if this was part of the constructor? I would prefer to expose the most minimal interface that still gets the job done.

Reply:
Good question - there would be no loss of functionality by moving this to the constructor. This method was created due to my preference to reduce any logic in the constructor itself. I'll update and replace with a constructor.